pkg/policyviolation/namespacedpv.go

```diff
@@ -8,138 +8,103 @@ import (
-		}
-
-		// compare the policyviolation spec for existing resource if present else
-		if reflect.DeepEqual(curPv.Spec, newPv.Spec) {
-			// if they are equal there has been no change so dont update the polivy violation
-			glog.V(3).Infof("namespaced policy violation '%s' spec did not change so not updating it", newPv.Spec.ToKey())
-			glog.V(4).Infof("namespaced policy violation spec %v did not change so not updating it", newPv.Spec)
-			continue
-		}
-
-		// set newPv name with curPv, as we are updating the resource itself
-		newPv.SetName(curPv.Name)
+func (nspv *namespacedPV) createPV(newPv *kyverno.NamespacedPolicyViolation) error {
+	var err error
+	glog.V(4).Infof("creating new policy violation for policy %s & resource %s/%s/%s", newPv.Spec.Policy, newPv.Spec.ResourceSpec.Kind, newPv.Spec.ResourceSpec.Namespace, newPv.Spec.ResourceSpec.Name)
+	obj, err := retryGetResource(nspv.dclient, newPv.Spec.ResourceSpec)
+	if err != nil {
+		return fmt.Errorf("failed to retry getting resource for policy violation %s/%s: %v", newPv.Name, newPv.Spec.Policy, err)
+	}
+	// set owner reference to resource
+	ownerRef := createOwnerReference(obj)
+	newPv.SetOwnerReferences([]metav1.OwnerReference{ownerRef})
 
-		// spec changed so update the policyviolation
-		glog.V(4).Infof("creating new policy violation for policy %s & resource %s", curPv.Spec.Policy, curPv.Spec.ResourceSpec.ToKey())
-		//TODO: using a generic name, but would it be helpful to have naming convention for policy violations
-		// as we can only have one policy violation for each (policy + resource) combination
-		if _, err = pvInterface.NamespacedPolicyViolations(namespace).Update(&newPv); err != nil {
-			return fmt.Errorf("failed to update namespaced policy violation: %v", err)
-		}
-		glog.Infof("namespaced policy violation updated for resource %s", newPv.Spec.ResourceSpec.ToKey())
+	// create resource
+	_, err = nspv.kyvernoInterface.NamespacedPolicyViolations(newPv.GetNamespace()).Create(newPv)
+	if err != nil {
+		glog.V(4).Infof("failed to create Cluster Policy Violation: %v", err)
+		return err
 	}
+	glog.Infof("policy violation created for resource %v", newPv.Spec.ResourceSpec)
 	return nil
 }
 
-func getExistingNamespacedPVIfAny(nspvLister kyvernolister.NamespacedPolicyViolationLister, newPv kyverno.NamespacedPolicyViolation) (kyverno.NamespacedPolicyViolation, error) {
-	// TODO(shuting): list pvs by labels
-	pvs, err := nspvLister.NamespacedPolicyViolations(newPv.GetNamespace()).List(labels.NewSelector())
-	if err != nil {
-		return kyverno.NamespacedPolicyViolation{}, fmt.Errorf("failed to list namespaced policy violations err: %v", err)
+func (nspv *namespacedPV) updatePV(newPv, oldPv *kyverno.NamespacedPolicyViolation) error {
+	var err error
+	// check if there is any update
+	if reflect.DeepEqual(newPv.Spec, oldPv.Spec) {
+		glog.V(4).Infof("policy violation spec %v did not change so not updating it", newPv.Spec)
+		return nil
 	}
+	// set name
+	newPv.SetName(oldPv.Name)
```

Comment:
Also need to set resource version if it's an update, https://github.com/nirmata/kyverno/blob/a5aa8669ff772134aa2b1dac92b9bb15c7b1e68c/pkg/policyviolation/namespacedpv.go#L102

Reply:
Corrected